### PR TITLE
add lowlevel thread support

### PR DESCRIPTION
**Pull request by rainers**
another extract of https://github.com/dlang/druntime/pull/2514

### CODE REVIEW

**Comment by CyberShadow** [comment]
Could the high-level thread code be refactored on top of the low-level code added here?

**Comment by rainers** [comment]
> Could the high-level thread code be refactored on top of the low-level code added here?

Possible, but with a couple of complications, because some high level variables are changed inbetween (e.g. m_hndl, m_isRunning). That would probably need a greater refactoring.

**Comment by CyberShadow** [comment]
If the symbols added here are to be public, then I think that would make this a situation that @andralex asked to be notified about.

**Comment by rainers** [comment]
I wouldn't mind keeping the symbols "private", i.e. undocumented, but I want to call them from the GC.

**Comment by jacob-carlborg** [comment]
> I wouldn't mind keeping the symbols "private", i.e. undocumented, but I want to call them from the GC.

The GC is not in the `core` top level package?

**Comment by rainers** [comment]
> The GC is not in the `core` top level package?

No, it has its own top level package `gc`.

**Comment by thewilsonator** [comment]
This needs another patch to weka-io/mecca

**Comment by thewilsonator** [comment]
That patch for weka is blocked on https://issues.dlang.org/show_bug.cgi?id=19761

File: src/core/thread.d [hunk 1]
```diff
@@ -1716,7 +1716,12 @@ private:
         return cast(Mutex)_locks[1].ptr;
     }
 
-    __gshared align(Mutex.alignof) void[__traits(classInstanceSize, Mutex)][2] _locks;
+    @property static Mutex lowLevelThreadslock() nothrow @nogc
+    {
+        return cast(Mutex)_locks[2].ptr;
+    }
+
+    __gshared align(Mutex.alignof) void[__traits(classInstanceSize, Mutex)][3] _locks;
```
Comment (ibuclaw) [inline src/core/thread.d]: Can this be its own private field to not interfere with #2411 (accessing this array causes crash with SPARC, SystemZ, PA-RISC, or any Solaris platforms).

Comment (ibuclaw) [reply src/core/thread.d]: From what I can tell from other comments, you will also avoid being blocked by weka too.

Comment (rainers) [reply src/core/thread.d]: Thanks for the hint, now using a separate lock.

File: src/core/thread.d [hunk 2]
```diff
@@ -5628,3 +5631,187 @@ version (Windows)
...
+
+    void initLowlevelThreads() nothrow @nogc
+    {
+        ll_lock[] = typeid(Mutex).initializer[];
+        lowlevelLock.__ctor();
+    }
+
+    void termLowlevelThreads() nothrow @nogc
+    {
+        lowlevelLock.__dtor();
```
Comment (ibuclaw) [inline src/core/thread.d]:
```
core/thread.d(5657): Error: destructor `core.sync.mutex.Mutex.~this` is not `nothrow`
core/thread.d(5655): Error: `nothrow` function `core.thread.termLowlevelThreads` may throw
```

**Comment by thewilsonator** [comment]
Why does CircleCI fail to install DMD?

**Comment by thewilsonator** [comment]
This otherwise looks go to me but CircleCI is required.

File: src/core/thread.d [hunk 3]
```diff
@@ -5628,3 +5631,187 @@ version (Windows)
...
     alias ThreadID = pthread_t;
+
+///////////////////////////////////////////////////////////////////////////////
+// lowlovel threading support
+private
+{
+    __gshared size_t ll_nThreads;
+    __gshared ThreadID* ll_pThreads;
+
+    __gshared align(Mutex.alignof) void[__traits(classInstanceSize, Mutex)] ll_lock;
```
Comment (ibuclaw) [inline src/core/thread.d]: Note: this is alignment is insufficient for targets with strict alignment requirements.

Comment (rainers) [reply src/core/thread.d]: Yes, will need to be updated in lock-step with Thread._locks.

**Pull request merged**